components/tikv_util/src/quota_limiter.rs

```diff
@@ -149,40 +196,65 @@ impl QuotaLimiter {
             .store(duration.0.as_nanos() as u64, Ordering::Relaxed);
     }
 
+    pub fn set_enable_auto_tune(&self, enable_auto_tune: bool) {
+        self.enable_auto_tune
+            .store(enable_auto_tune, Ordering::Relaxed);
+    }
+
+    pub fn foreground_cputime_limiter(&self) -> f64 {
+        self.foreground_cputime_limiter.speed_limit()
+    }
+
+    pub fn background_cputime_limiter(&self) -> f64 {
+        self.background_cputime_limiter.speed_limit()
+    }
+
     fn max_delay_duration(&self) -> Duration {
         Duration::from_nanos(self.max_delay_duration.load(Ordering::Relaxed))
     }
 
+    pub fn supports_auto_tune(&self) -> bool {
+        self.enable_auto_tune.load(Ordering::Relaxed)
+    }
+
     // To generate a sampler.
     pub fn new_sample(&self) -> Sample {
         Sample {
             read_bytes: 0,
             write_bytes: 0,
             cpu_time: Duration::ZERO,
-            enable_cpu_limit: !self.cputime_limiter.speed_limit().is_infinite(),
+            enable_cpu_limit: !self.foreground_cputime_limiter.speed_limit().is_infinite()
+                || !self.background_cputime_limiter.speed_limit().is_infinite(),
         }
     }
 
     // To consume a sampler and return delayed duration.
     // If the sampler is null, the speed limiter will just return ZERO.
-    pub async fn async_consume(&self, sample: Sample) -> Duration {
+    pub async fn consume_sample(&self, sample: Sample, is_foreground: bool) -> Duration {
+        let mut cpu_limiter = self.background_cputime_limiter.clone();
+        let mut write_limiter = self.background_write_bandwidth_limiter.clone();
+        let mut read_limiter = self.background_read_bandwidth_limiter.clone();
+
+        if is_foreground {
+            cpu_limiter = self.foreground_cputime_limiter.clone();
+            write_limiter = self.foreground_write_bandwidth_limiter.clone();
+            read_limiter = self.foreground_read_bandwidth_limiter.clone();
+        }
```

Comment:
Does this work? `clone` can easily become a bottleneck when `consume_sample` is frequently called.
```suggestion
        let (cpu, write, read) = if is_foreground {
            (&self.foreground_cputime_limiter, ...)
        } else { (&self.background_cputime_limiter, ...) }
```

Reply:
> Does this work? `clone` can easily become a bottleneck when `consume_sample` is frequently called.

Good suggestion.